### PR TITLE
Adds more robust timer duration parsing

### LINKED ISSUE #19454: Input validation for Trainer's `min_time` and `max_time` when passed as string
### Description & Motivation

The argument to the `Trainer`s min_time is expected to be a string of the form /(\d\d):(\d\d):(\d\d):(\d\d)/. One way in which parsing can fail (simple example: `--min_time 60`) throws an `IndexError` and provides little context for what is wrong.

### Pitch

This is an uninformative error message, and managed to briefly stump me and a colleague. It should be caught and converted into an informative one that explains the expected format.

This should probably be done at the API level rather than becoming something specific to CLI parsing. That is, since one can provide a string to this API, it should catch this error and instead raise an error with a name like `TimeParsingError` or something of that nature.

### Alternatives

Use `datetime.timedelta.strptime` or other standard-library elements to do the parsing, and catch the `ValueError` and raise something more informative.

### Additional context

I am working on PyTorch-Lightning >=1.7.0,<2.0.0 because I haven't had the chance to migrate to `LightningCLI` and such, but a quick perusal suggests this is still an issue.

cc @borda @carmocca @awaelchli

### PR DESCRIPTION
**Pull request by kylebgorman**
## What does this PR do?

Failure to provide the timer duration string (e.g., via `--max_time`) in the appropriate format currently results in an uninformative IndexError. This uses a simple regex (and the built-in `re`) for more robust input validation and instead throws a misconfiguration exception.

I experimented with `datetime.strptime` for this, but it is not well-suited to the job because it is designed for absolute times, not deltas, and therefore does not support the idea of "0 days".

Fixes #19454.

I am submitting as is, but reviewers, feel free to recommend testing. 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19513.org.readthedocs.build/en/19513/

### CODE REVIEW

**Review by awaelchli — COMMENTED** [review]
@kylebgorman This is already great!

Would you like to try adding a unit test for this? 
It would go into the file `tests/tests_pytorch/callbacks/test_timer.py`. You can use 

```
with pytest.raises(Misconfiguration, match="Expected DD:HH:MM:SS format"):
    Timer(...)
```
to test your code.

Otherwise I'm happy to help or do it.

File: src/lightning/pytorch/callbacks/timer.py [hunk 1]
```diff
@@ -86,9 +89,17 @@ def __init__(
     ) -> None:
         super().__init__()
         if isinstance(duration, str):
-            dhms = duration.strip().split(":")
-            dhms = [int(i) for i in dhms]
-            duration = timedelta(days=dhms[0], hours=dhms[1], minutes=dhms[2], seconds=dhms[3])
+            datematch = re.fullmatch(r"(\d+):(\d\d):(\d\d):(\d\d)", duration.strip())
```
Comment (awaelchli) [inline src/lightning/pytorch/callbacks/timer.py:92]:
```suggestion
            duration_match = re.fullmatch(r"(\d+):(\d\d):(\d\d):(\d\d)", duration.strip())
```

**Comment by kylebgorman** [comment]
I greatly appreciate the suggestions, particularly about wording of the exception. I'll add a unit test shortly.

**Comment by kylebgorman** [comment]
> That's great, thanks a lot for adding the test! Feel free to also add an entry in the "Added" section of the changelog in src/lightning/pytorch/CHANGELOG.md

Done.

**Review by carmocca — APPROVED** [review]

**Pull request merged**